File: cpp/LCS.cpp
#include <iostream>
#include <string>

using namespace std;

int lcs(string X, string Y, int m, int n)
{
  int L[m+1][n+1];
  for(int i = 0; i <= m; i++) {
    for(int j = 0; j <= n; j++) {
      if(i == 0 || j == 0)
        L[i][j] = 0;
      else if(X[i - 1] == Y[j - 1])
        L[i][j] = L[i - 1][j - 1] + 1;
      else
        L[i][j] = max(L[i - 1][j], L[i][j-1]);
    }
  }
return L[m][n];
}

int main()
{
  string X = "AGGTAB";
  string Y = "GXTXAYB";

  int m = X.length();
  int n = Y.length();

  cout << "length of lcs is " << lcs(X, Y, m, n) << end1;

  return 0;
}

File: ocaml/LCS.ml
let lcs str1 str2 =
  let m = String.length str1 in
  let n = String.length str2 in 
  let dp = Array.make_matrix (m + 1) (n + 1) 0 in

  for i = 1 to m do
    for j = 1 to n do
      if String.get str1 (i - 1) = String.get str2 (j - 1) then
        dp.(i).(j) <- dp.(i - 1).(j-1) + 1
      else
        dp.(i).(j) <- max dp.(i-1).(j) dp.(i).(j-1)
    done
  done
  dp.(m).(n)

let () = 
  let str1 = "AGGTAB" in 
  let str2 = "GXTXAYB" in
  let lcs_length = lcs str1 str2 in
  Printf.printf "Length of LCS is %d\n" lcs_length 
